hydra_base/lib/template.py

```diff
@@ -93,72 +102,95 @@ def get_types_by_attr(resource, template_id=None):
+        Search for an attribute with the given name and dimension_id.
         If such an attribute does not exist, create one.
     """
 
-    attr = db.DBSession.query(Attr).filter(Attr.name==name,
-                                           Attr.dimension==dimension).first()
+    attr = db.DBSession.query(Attr).filter(Attr.name==name, Attr.dimension_id==dimension_id).first()
 
     if attr is None:
+        # In this case the attr does not exists so we must create it
         attr         = Attr()
-        attr.dimension = dimension
+        attr.dimension_id = dimension_id
         attr.name  = name
 
-        log.info("Attribute not found, creating new attribute: name:%s, dimen:%s",
-                    attr.name, attr.dimension)
+        log.debug("Attribute not found, creating new attribute: name:%s, dimen:%s",
+                    attr.name, attr.dimension_id)
 
         db.DBSession.add(attr)
 
     return attr
 
 def parse_xml_attribute(attribute):
+    dimension_i = None
 
-    if attribute.find('dimension') is not None:
+    if attribute.find('dimension') is not None and attribute.find('dimension').text is not None:
+        # Found the dimension
         dimension_name = attribute.find('dimension').text
-        if dimension_name is not None:
-            if dimension_name.lower() in ('dimensionless', ''):
-                dimension_name = 'dimensionless'
+        if dimension_name.lower() in ('dimensionless', ''):
+            dimension_name = 'dimensionless'
         dimension_i = units.get_dimension_by_name(dimension_name.strip())
 
-    elif attribute.find('unit') is not None:
-        if attribute.find('unit').text is not None:
-            dimension_i = units.get_unit_dimension(attribute.find('unit').text)
+    elif attribute.find('unit') is not None and attribute.find('unit').text is not None:
+        # Found the unit
+        unit = attribute.find('unit').text
+        unit_id = units.get_unit_by_abbreviation(unit).id
+        #dimension_i = units.get_dimension_by_unit_id(unit_id)
+        dimension_i = units.get_dimension_by_unit_id(unit_id, do_accept_unit_id_none=True)
+
     else:
-        raise HydraError("An attribute must have a unit or dimension. %s has neither"%(attribute_i.name))
+        # No unit and no dimension. Let's get/create it anyway
+        pass
+    # if dimension_i is None:
+    #     raise HydraError("An attribute must have a unit or dimension. %s has neither"%(attribute_i.name))
 
-    name      = attribute.find('name').text.strip()
+    name = attribute.find('name').text.strip()
 
-    attr = _get_attr_by_name_and_dimension(name, dimension_i.name)
+    if dimension_i is None:
+        attr = _get_attr_by_name_and_dimension(name, None)
+    else:
+        attr = _get_attr_by_name_and_dimension(name, dimension_i.id)
 
     db.DBSession.flush()
 
     return attr
 
+def _exists_key_in_object(key, obj):
+    my_dict = JSONObject(obj)
+    try:
```

Comment:
In python you can just use 'in'. Please use that instead of this.
Ex:
```pyhton
if key in obj:
    pass
```

Alternatively, you can say:
```<python>
if obj.get(key) is not None:
    #do something
```

Reply:
Steve, i know the comparisons you talked about, but in some cases this comparisons does not work.

For example some objects are not iterable and the comparison itself emits an exception, like in all the cases for which I used this function:

```python
>       if key in obj:
E       TypeError: argument of type 'TypeAttr' is not iterable
```
Or in the get function is not defined:
```python
>       if obj.get(key) is not None:
E       AttributeError: 'TypeAttr' object has no attribute 'get'
```

I just used the **EAFP** way that in Python is widely used.
I am sure that you know this url.

https://docs.python.org/3/glossary.html#term-eafp

Therefore we cannot avoid this functionality or the test will not pass, with all the consequences

Reply:
Otherwise we must do the object iterable/ add the get method.